nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy

```diff
@@ -0,0 +1,159 @@
+            }
+            if(truststore.startsWith("./")){
+                truststore = truststore.replace("./",nifiInstallDir+"/")
+            }
+
+            sslContext = createSslContext(
+                    keystore.trim(),
+                    keystorePassword.trim().toCharArray(),
+                    keystoreType.trim(),
+                    truststore.trim(),
+                    truststorePassword.trim().toCharArray(),
+                    truststoreType.trim(),
+                    "TLS");
+        }
+
+        final ClientConfig config = new DefaultClientConfig();
+
+        if (sslContext != null) {
+            config.getProperties().put(HTTPSProperties.PROPERTY_HTTPS_PROPERTIES,new HTTPSProperties(new NiFiHostnameVerifier(), sslContext));
+        }
+
+        return  Client.create(config);
+
+    }
+
+
+    private static SSLContext createSslContext(
+            final String keystore, final char[] keystorePasswd, final String keystoreType,
+            final String truststore, final char[] truststorePasswd, final String truststoreType,
+            final String protocol)
+            throws KeyStoreException, IOException, NoSuchAlgorithmException, CertificateException,
+                    UnrecoverableKeyException, KeyManagementException {
+
+        // prepare the keystore
+        final KeyStore keyStore = KeyStore.getInstance(keystoreType);
+        final InputStream keyStoreStream = new FileInputStream(keystore)
+            keyStore.load(keyStoreStream, keystorePasswd);
+
+
+        final KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        keyManagerFactory.init(keyStore, keystorePasswd);
+
+        // prepare the truststore
+        final KeyStore trustStore = KeyStore.getInstance(truststoreType);
+        final InputStream trustStoreStream = new FileInputStream(truststore)
+        trustStore.load(trustStoreStream, truststorePasswd);
+
+        final TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        trustManagerFactory.init(trustStore);
+
+        // initialize the ssl context
+        final SSLContext sslContext = SSLContext.getInstance(protocol);
+        sslContext.init(keyManagerFactory.getKeyManagers(), trustManagerFactory.getTrustManagers(), new SecureRandom());
+        return sslContext;
+    }
+
+    private static class NiFiHostnameVerifier implements HostnameVerifier {
+
+        @Override
+        public boolean verify(final String hostname, final SSLSession ssls) {
+            try {
+                for (final Certificate peerCertificate : ssls.getPeerCertificates()) {
+                    if (peerCertificate instanceof X509Certificate) {
+                        final X509Certificate x509Cert = (X509Certificate) peerCertificate;
+                        final List<String> subjectAltNames = getSubjectAlternativeNames(x509Cert);
```

Comment:
The hostname should first be compared to the hostname in the subject DN of the certificate before resorting to the SANs. There are two scenarios where this would return incorrectly -- the cert hostname matches but no SANs match (false negative) or the cert has no SANs (false negative). See [`X509Certificate.getSubjectX500Principal()`](https://docs.oracle.com/javase/7/docs/api/java/security/cert/X509Certificate.html#getSubjectX500Principal()) to access the subject DN directly and [`X500Principal.getName()`](https://docs.oracle.com/javase/7/docs/api/javax/security/auth/x500/X500Principal.html#getName()) to extract the hostname (`CN` value) from the principal for comparison. If this comparison fails, then check the SANs.